### ISSUE TITLE
New questions about codes

### ISSUE THREAD

**Issue by SMZCC** [liulj13/pyMDNet-VOT2OTB#2]
#####  Questions about data_prov.py
  - When determine how many pos and neg samples per frame to sample , why use formulas like these:
    + line 54 `n_pos = (self.batch_pos - len(pos_regions)) // (self.batch_frames - i)`
    + line 55 `n_neg = (self.batch_neg - len(neg_regions)) // (self.batch_frames - i)`
    + Aren't the results of formulas above constants? why don't we use `self.batch_neg // self.batch_frames`  and `self.batch_pos // self.batch_frames` instead?
  - line 73, `regions = regions.astype('float32') - 128` , at the last of the extract regions, why should the regions minus 128?

##### Questions about sample_generator.py
  - line 14, `factor = 2`, now that we have known how many samples to sample per frame, why should we multiply the `remain` with the `factor` ? what's the meaning of the factor ?`samples_ = generator(bbox, remain*factor)`
- line 15, `while remain > 0 and factor < 16:`I can understand the condition of `remain > 0` , however ,why should the factor < 16? Can the factor smaller than the other any numbers? for example 13,12 etc.
- line 53, `sample = np.array([bb[0]+bb[2]/2, bb[1]+bb[3]/2, bb[2], bb[3]], dtype='float32')`, when sampling,  why should we make the boundingbox of right-bottom corner of the groundthruth as the blueprint rather than using groundtruth itself ?
- when sampling , `samples[:,2:] *= self.scale_f ** np.clip(0.5*np.random.randn(n,1),-1,1) `, why should we use the form of  `w * aspect_f ^ some_number` rather than directly use ` w * aspect_f` ?
- line 70 ~  76, how does them represent the concept of 'whole', why do them equal to the union of  two ways above which are 'gaussian' and 'uniform' ? 
 ```python
        elif self.type=='whole':
            m = int(2*np.sqrt(n))
            xy = np.dstack(np.meshgrid(np.linspace(0,1,m),np.linspace(0,1,m))).reshape(-1,2)
            xy = np.random.permutation(xy)[:n]
            samples[:,:2] = bb[2:]/2 + xy * (self.img_size-bb[2:]/2-1)  # img_size: (w, h)
            #samples[:,:2] = bb[2:]/2 + np.random.rand(n,2) * (self.img_size-bb[2:]/2-1)
            samples[:,2:] *= self.scale_f ** (np.random.rand(n,1)*2-1)
 ```
 - line 81, 
      ```python
        if self.valid:
            samples[:,:2] = np.clip(samples[:,:2], samples[:,2:]/2, self.img_size-samples[:,2:]/2-1)
        else:
            samples[:,:2] = np.clip(samples[:,:2], 0, self.img_size)
      ```   
      + why should the coordinates of valid samples  be limited to [ samples[:,2:]/2, self.img_size-samples[:,2:]/2-1] rather than [ samples[:,2:]/2, self.img_size-1] or any other intervals?
      + and there is a relative question
      + line 86, `samples[:,:2] -= samples[:,2:]/2`, now that ,at the last of the `__call__` of `sample_generator` ,we will make the `w` and `h` minus `half_w`, `half_h`, why should we add `half_w`, `half_h` at the beginning? `sample = np.array([bb[0]+bb[2]/2, bb[1]+bb[3]/2, bb[2], bb[3]], dtype='float32')`

##### Questions about utils.py
  - line 34
    ```python
     if padding > 0:
        pad_w = padding * w/img_size
        pad_h = padding * h/img_size
        half_w += pad_w
        half_h += pad_h
    ```
    + when cropping the image ,why do we need a padding ? What's the function of padding ?
        + can't we use the bounding-boxes from samples directly to crop the image?
        + why should the padding be processed like this `pad_w = padding * w/img_size`
 - line 61, `cropped = 128 * np.ones((max_y-min_y, max_x-min_x, 3), dtype='uint8')`, here,why not use `cropeed = np.zeros((max_y-min_y, max_x-min_x), dtype='uint8')` instead? Generally, don't we use the zero function ?
- line 62 , ` cropped[min_y_val-min_y:max_y_val-min_y, min_x_val-min_x:max_x_val-min_x, :] \
            = img[min_y_val:max_y_val, min_x_val:max_x_val, :]` why the indices of cropped are `[min_y_val-min_y:max_y_val-min_y, min_x_val-min_x:max_x_val-min_x, :]`

**Comment by liulj13** [comment]
In the data_prov.py:

For the first question, in our case the self.batch_pos is 32 and self.batch_frames is 8, so the n_pos is a constant(32/8=4), but if the self.batch_pos is 30, the n_pos will be (30/8=3,27/7=3,24/6=4,20/5=4,16/4=4,12/3=4,8/2=4,4/1=4), and the n_pos is not a constant.

For the minus 128 process, substracting the median value is a common preprocessing of many problems in deep learning based computer vision problems. The pixel value is from 0-255, and the 128 is the median value. Since computing the average pixel value of the whole dataset is complicated and cost too much time, so we substract 128 instead.

In the sample_generator.py:

For the first question, the factor is to ensure that we can sample enough samples that satisfy the overlap constraint. We want to generate n samples that satisfy the overlap constraint, but the sampling process does not ensure that the generated samples can satisfy the overlap constraint, so we samples n*factor samples and save the good samples. And we have do it multiple times to ensure we sample enough samples. The factor is 2->4->8->16, when we do not have enough samples that satisfy the overlap constraint, the solution is to sample more and more samples. So the 16 is an experimental number, it can be 32,64,..., the only difference is the computation time.

For the line 53, since the gt is using the (left_top_x, left_top_y, w, h), but the sampling process is using the (center_x, center_y, w, h), so we have to change the format. The reason is that using the center_x and center_y is more convenient to sample translation as line 63.

When sampling, the scale_f is 1.2, and the numpy.clip makes the output to [-1,1], so the scale variation is 1.2^[-1,1] which is from 0.8333 to 1.2. We use this approach to create many different scales.

The whole means generating samples from the whole image, and it is used to sample negative samples at the initial frame.

For the line 81, this is because sometimes we samples the sample which overgos the boundary of the whole image, so we have to fix it. The [samples[:,2:]/2, self.img_size-samples[:,2:]/2-1] can ensure that the sample will not overgo the boundary of the whole image but the [samples[:,2:]/2, self.img_size-1] cannot, because when the center is near the boundary the size of sample should be much smaller.

For line 86, it is because the sampling process use (center_x, center_y, w, h) and after sampling, we have to recover the format.

In the utils.py:

For the line 34, the padding is to include more information around the sampled bounding boxes, so we use pad_w = padding * w/img_size to calculate the 16 in w can be which numer in 107.

For the line 61, since some padding point do not have values, we give them the value of 128, which is the median value of 0-255, and this is better than give them the value of zero.

For the line 62, since min_y can be negative, so we cannot use line 53 cropped = img[min_y:max_y, min_x:max_x, :] directly.

Good Luck~

**Comment by SMZCC** [comment]
#### Wow , you're so great, to me , thx
##### A further question
  - You just mentioned above
    > The reason is that using the center_x and center_y is **more convenient** to sample translation as line 63.
    + Now that the gt is represented as `(left_top_x, left_top_y, w, h)`  and  in the sampling process 
       we use `(center_x, center_y, w, h)`.The two ways  are both representing a dot with the 
       corresponding width and height, then, why could we say  `using the center_x and center_y is more convenient  to sample translation` ?
#### Best regards

**Comment by liulj13** [comment]
Sorry for the late response, I have been overwhelmingly busy during this time~

For the first problem, when sampling, we first sample the center, then sample the width and length of the bounding boxes. But if we use the left_top_x and left_top_y, then sample the width and length of the bounding boxes, the candidates we get will all have the same left_top_x and left_top_y. The reason may not be convenient, but beacause for visual tracking, the scale variation may only change the width and length and keep the center position unchanged(left_top_x and left_top_y is changed), so we use the center for the sampling process to model the scale variation process.

For the second problem, the valid is false by default and is true for sampling testing candidates. When training, we can use the invalid samples(which overgo the boundary of the whole image), but when testing, we only use the valid samples(which do not overgo the boundary of the whole image). The reason is because the training process may not be so strict, and the testing process we cannot tolerate an invalid sample becoming tracking result.

For the third problem, since in line 73, we use xy = np.random.permutation(xy)[:n] , so the number of 2 can be changed, and actually, you can debug to this place and see what happens!

For the fourth problem, during training, the Accuracy class is not used, and the Precision class is used. Both of these two class can measure the current performance of the model during training. You can print these two criteria to see their values or debug into them~

**Comment by SMZCC** [comment]
- > Sorry for the late response, I have been overwhelmingly busy during this time~

##### That's alright. :smile: 
- > For the first problem ....
   For the second problem

I have known, thx. :+1: 

- > For the third problem, since in line 73, we use xy = np.random.permutation(xy)[:n] , so the number of 2 can be changed, and actually, you can debug to this place and see what happens!

Now, I think that only changes the number of candidate samples and the number ,2,  is arbitrary.Am I right? :thinking: 

- > Both of these two class can measure the current performance of the model during training

Is that to say we actually only need one Accuracy class and the another is optional and we can just delete it ? :thinking: 

- Did you get the results mentioned in the paper of MDNet? I find it so hard to train the model(I only have a notebook of 4G GPU and i5 7th generation),even if I find the [author's implemention and using the parameters the author provided](https://github.com/HyeonseobNam/py-MDNet), the results are far away from the paper :cry: I think I will die, I have no much time..... :sob: 

- I also find the [bendmarch of python](https://github.com/jwlim/tracker_benchmark), however, how to use it to draw precision of location? I find it only draws the success rate of overlap  and the number in the bracket is larger than 1 while in the papers , the numbers are between 0 and 1, why?  :thinking: 
- When using the benchmark, can we only give the benchmark results rather than embedding the tracker in it ? Because when I embedding the tracker in it , it always encounter some strange problems which almostly makes me mad :angry:  
##### Best regards

**Comment by liulj13** [comment]
For the first two questions, you are right~

For your hardware, you can use the matlab version code provided by the authors, it requires only 2 GB GPU: [https://github.com/HyeonseobNam/MDNet](here)

And the benchmark also have a matlab version which is better I think:
[http://cvlab.hanyang.ac.kr/tracker_benchmark/index.html](here)

Yes, you can give the benchmark results rather than embedding the tracker in it.

**Comment by SMZCC** [comment]
> it requires only 2 GB GPU

Why the matlab only needs 2G GPU?

**Comment by liulj13** [comment]
> Why the matlab only needs 2G GPU?
Because it is an old version and it is slower, actually you can also reduce the batchsize in the pytorch version code to use less GPU with a slower running speed~

**Comment by SMZCC** [comment]
>  reduce the batchsize 

If I reduce the batchsize , will the result be affected ? :thinking:

**Comment by liulj13** [comment]
It will be a little different but the overall performance will not degrade too much, actually each time you run the code you will get a result with a little difference.
